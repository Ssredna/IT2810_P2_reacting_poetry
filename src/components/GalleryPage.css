.GalleryPage-container {
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
  justify-content: space-evenly;
  gap: 2rem;
  padding-bottom: 1rem;
}
.GalleryPage-title {
  text-align: center;
  margin-top: 0.6rem;
  margin-bottom: 0.3rem;
}

.GalleryPage-slider {
  -webkit-appearance: none;
  height: 0.75rem;
  background: #d3d3d3;
  border-radius: 10px;
  outline: none;
  opacity: 0.7;
  -webkit-transition: 0.2s;
  transition: opacity 0.2s;
}

.GalleryPage-slider:hover {
  opacity: 1;
  cursor: pointer;
}

.GalleryPage-slider::-webkit-slider-thumb {
  -webkit-appearance: none;
  appearance: none;
  width: 1.4rem;
  height: 1.4rem;
  background: #4c4344;
  cursor: pointer;
  border-radius: 50%;
}

.GalleryPage-slider::-moz-range-thumb {
  width: 1.4rem;
  height: 1.4rem;
  background: #4c4344;
  cursor: pointer;
  border-radius: 50%;
}

.GalleryPage-figure-amount-meta {
  font-size: 1rem;
  margin-top: 0.4rem;
  margin-bottom: 0.2rem;
}

.GalleryPage-figure-amount {
  font-size: 1.3rem;
}

.GalleryPage-actions {
  display: flex;
  flex-direction: column-reverse;
  justify-content: space-around;
  align-items: center;
  margin-bottom: 2rem;
}
.GalleryPage-dropbtn {
  padding: 0.8rem;
  margin-top: 0.4rem;
  font-size: 1rem;
  border: none;
  cursor: pointer;
  border-radius: 5px;
  transition: 0.2s;
  font-family: 'Roboto Slab', serif;
}
.GalleryPage-dropbtn:hover {
  box-shadow: 4px 8px 16px 4px rgba(0, 0, 0, 0.2);
}

.GalleryPage-dropdown {
  position: relative;
  display: inline-block;
}

.GalleryPage-dropdown-content {
  display: none;
  position: absolute;
  border-radius: 3px;
  top: 3.2rem;
  left: -0.5rem;
  background-color: #f5eeee;
  min-width: 11rem;
  box-shadow: 4px 8px 16px 4px rgba(0, 0, 0, 0.2);
  z-index: 1;
  transition: 0.3s;
}

.GalleryPage-dropdown-content button {
  color: black;
  padding: 1rem 1rem;
  display: block;
  border: none;
  width: 100%;
  font-size: 1rem;
  font-family: 'Roboto Slab', serif;
}

.GalleryPage-dropdown-content button:hover {
  background-color: #a59092;
  cursor: pointer;
}

.GalleryPage-dropdown-content button:focus {
  outline: none;
}

.GalleryPage-dropdown-content button:active {
  background-color: #7a5e61;
}

.GalleryPage-dropdown:hover .GalleryPage-dropdown-content {
  display: block;
  cursor: pointer;
}

.GalleryPage-dropdown:hover .GalleryPage-dropbtn {
  background-color: #d69696;
}

.GalleryPage-complementary-btn {
  margin: 0.3rem;
  padding: 0.8rem;
  font-size: 1.1rem;
  border-radius: 10px;
  border: none;
  background: linear-gradient(-45deg, #e2e2e2, #a79191, #c9bfbf);
  font-family: 'Roboto Slab', serif;
  transition: 0.5s;
  outline: none;
}

.GalleryPage-complementary-btn-fancy {
  background: linear-gradient(
    -45deg,
    #fcc5c5,
    #c0ffd5,
    #bbedd7,
    #edbbd1,
    #add8e6
  );
  animation: fancyMode 6s infinite linear;
}

@keyframes fancyMode {
  0% {
    transform: scale(1);
  }
  3% {
    transform: scale(0.86);
  }
  5% {
    transform: scale(1.08) rotate(7deg);
  }
  10% {
    transform: scale(0.9);
  }
  15% {
    transform: scale(1.06) rotate(-7deg);
  }
  20% {
    transform: scale(0.95);
  }
  25% {
    transform: scale(1.03);
  }
  50% {
    transform: translate(0.4rem) rotateZ(6deg);
  }
  60% {
    transform: translate(-0.3rem);
  }
  70% {
    transform: translate(0.2rem) rotateZ(-6deg);
  }
  95% {
    transform: translate(-0.12rem);
  }

  100% {
    transform: scale(1);
  }
}

@media screen and (min-width: 37.5em) {
  .GalleryPage-actions {
    flex-direction: row;
  }
  .GalleryPage-dropbtn {
    padding: 1rem;
    font-size: 1.2rem;
  }
  .GalleryPage-slider::-webkit-slider-thumb {
    width: 2rem;
    height: 2rem;
  }
  .GalleryPage-slider::-moz-range-thumb {
    width: 2rem;
    height: 2rem;
  }

  .GalleryPage-figure-amount-meta {
    font-size: 1.2rem;
  }

  .GalleryPage-figure-amount {
    font-size: 1.5rem;
  }
}

@media screen and (min-width: 75em) {
  .GalleryPage-complementary-btn:hover {
    background: rgb(156, 128, 109);
    cursor: pointer;
    background: linear-gradient(
      -145deg,
      #fcc5c5,
      #c0ffd5,
      #bbedd7,
      #edbbd1,
      #add8e6
    );
  }

  .GalleryPage-complementary-btn-fancy:hover {
    background: linear-gradient(
      -125deg,
      #fcc5c5,
      #c0ffd5,
      #bbedd7,
      #edbbd1,
      #add8e6
    );
  }
  .GalleryPage-dropbtn {
    padding: 1.1rem;
    font-size: 1.3rem;
  }
}
