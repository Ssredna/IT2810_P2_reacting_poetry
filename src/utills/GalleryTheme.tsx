export const ColorTheme: AllColorThemesInterface = {
  theme1: {
    color1: '#fcc5c5',
    color2: '#D4AD9B',
    complementary: '#c0ffd5',
    name: {
      english: 'Pink',
      norwegian: 'Rosa',
    },
  },
  theme2: {
    color1: '#BBEDD7',
    color2: '#7fddb3',
    complementary: '#edbbd1',
    name: {
      english: 'Green',
      norwegian: 'Grønn',
    },
  },
  theme3: {
    color1: '#add8e6',
    color2: '#72bcd4',
    complementary: '#e6d8ad',
    name: {
      english: 'Blue',
      norwegian: 'Blå',
    },
  },
};

interface AllColorThemesInterface {
  theme1: ColorThemeInterface;
  theme2: ColorThemeInterface;
  theme3: ColorThemeInterface;
}

export interface ColorThemeInterface {
  color1: string;
  color2: string;
  complementary: string;
  figureColor?: string;
  name: languageOptions;
}

export interface languageOptions {
  english: string;
  norwegian: string;
}

export enum Languages {
  EN = 'english',
  NO = 'norwegian',
}
