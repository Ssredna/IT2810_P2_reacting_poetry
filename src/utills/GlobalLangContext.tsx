import React from 'react';

/* Oppretter en Context for Global Language */
const GlobalLangContext = React.createContext('');
/* Eksporterer Context Provider som den som eier context skal bruke */
export const GlobalLangProvider = GlobalLangContext.Provider;

/* Eksporterer context som forbrukerne bruker med useContext
-hooken */
export default GlobalLangContext;

/* Translations const!  */
export const translations = {
  english: {
    gallery: 'Gallery',
    home: 'Home',
    theme: 'Theme',
    figAmount: 'Number of figures',
    fancyMode: 'Fancymode',
    off: 'off',
    on: 'on',
    close: 'Close',
    createdBy: 'Created by',
    project: 'project',
    welcomeText: 'Welcome to',
    toGallery: 'To gallery',
    noFavorites: 'You do not have any favorites yet!',
    galleryLoading: 'Gallery is loading...',
  },
  norwegian: {
    gallery: 'Galleri',
    home: 'Hjem',
    theme: 'Tema',
    figAmount: 'Antall figurer',
    fancyMode: 'Fancymodus',
    off: 'av',
    on: 'p√•',
    close: 'lukk',
    createdBy: 'Laget av',
    project: 'prosjekt',
    welcomeText: 'Velkommen til',
    toGallery: 'Til galleriet',
    noFavorites: 'Du har ikke lagt til noen favoritter enda!',
    galleryLoading: 'Laster inn galleri...',
  },
};
