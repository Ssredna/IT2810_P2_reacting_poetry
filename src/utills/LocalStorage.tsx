import { useState } from 'react';

export enum StorageType {
  LocalStorage = 'localStorage',
  SessionStorage = 'sessionStorage',
}

export const useWebStorage = (key, initialValue, storageType: StorageType) => {
  const [storedValue, setStoredValue] = useState(() => {
    try {
      const item = window[storageType].getItem(key);
      return item ? JSON.parse(item) : initialValue;
    } catch (err) {
      console.error(err);
      return initialValue;
    }
  });

  const setValue = (value) => {
    try {
      const valueToStore =
        value instanceof Function ? value(storedValue) : value;
      setStoredValue(valueToStore);
      window[storageType].setItem(key, JSON.stringify(valueToStore));
    } catch (err) {
      console.error(err);
    }
  };

  return [storedValue, setValue];
};

export default useWebStorage;
