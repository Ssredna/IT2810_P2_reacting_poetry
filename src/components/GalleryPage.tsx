import React, {
  useState,
  MouseEvent,
  TouchEvent,
  useEffect,
  useContext,
} from 'react';
import '../ColorPalette.css';
import '../Container.css';
import './GalleryPage.css';
import '../General.css';
import '../assets/css/all.css';
import GalleryItem from './GalleryItem';
import { useWebStorage, StorageType } from '../utills/WebStorage';
import { ColorTheme } from '../utills/GalleryTheme';
import GlobalLangContext, { translations } from '../utills/GlobalLangContext';

const mp3s: NodeRequire[] = [];
mp3s[0] = require('../assets/testing.mp3');
mp3s[1] = require('../assets/testing.mp3');
mp3s[2] = require('../assets/testing.mp3');
mp3s[3] = require('../assets/testing.mp3');
mp3s[4] = require('../assets/testing.mp3');
mp3s[5] = require('../assets/testing.mp3');
mp3s[6] = require('../assets/testing.mp3');
mp3s[7] = require('../assets/testing.mp3');
mp3s[8] = require('../assets/testing.mp3');
mp3s[9] = require('../assets/testing.mp3');

interface Poem {
  title: string;
  author: string;
  lines: string[];
  isFavorite?: boolean;
}

const GalleryPage = () => {
  const lang = useContext(GlobalLangContext);

  const [colorTheme, setColorTheme] = useWebStorage(
    'Theme',
    ColorTheme.theme1,
    StorageType.LocalStorage
  );
  const [favoritePoems, setFavoritePoems] = useWebStorage(
    'FavoritePoems',
    [],
    StorageType.SessionStorage
  );
  const [figureAmount, setFigureAmount] = useWebStorage(
    'FigureAmount',
    5,
    StorageType.SessionStorage
  );

  const [complementaryTheme, setComplementaryTheme] = useWebStorage(
    'ComplementaryTheme',
    false,
    StorageType.SessionStorage
  );

  const [poems, setPoems] = useState<Poem[]>([]);
  const [loading, setLoading] = useState(false);
  const [showFavorites, setShowFavorites] = useState(false);

  useEffect(() => {
    setLoading(true);
    fetch('https://poetrydb.org/linecount,poemcount/8;10')
      .then((res: Response) => res.json())
      .then((poemData: Poem[]) => setPoems(poemData))
      .then(() => setLoading(false));
  }, []);

  const updateFigureAmount = (
    event: MouseEvent<HTMLInputElement> | TouchEvent<HTMLInputElement>
  ) => {
    const target = event.target as HTMLInputElement;
    setFigureAmount(target.valueAsNumber);
  };

  const onComplementaryToggle = () => {
    setComplementaryTheme(!complementaryTheme);
  };

  const onFavoriteClick = () => {
    setShowFavorites(!showFavorites);
  };

  const addFavoritePoem = (poem: Poem) => {
    poem.isFavorite = poem.isFavorite ? false : true;
    const currentPoemIndex = favoritePoems.indexOf(poem);
    if (currentPoemIndex === -1) {
      setFavoritePoems([...favoritePoems, poem]);
    } else {
      let tmpFavoritePoems = [...favoritePoems];
      tmpFavoritePoems[currentPoemIndex] = poem;
      setFavoritePoems(tmpFavoritePoems);
    }
  };

  return (
    <div>
      {loading ? (
        <div className="GalleryPage-loading-section">
          <p className="GalleryPage-loading-text">Gallery is loading...</p>
        </div>
      ) : (
        <div className="General-container">
          <h1 className="GalleryPage-title">{translations[lang].gallery}</h1>
          <div className="GalleryPage-actions">
            <div className="GalleryPage-dropdown">
              <button
                className="GalleryPage-dropbtn"
                style={{ background: colorTheme.color1 }}
              >
                {translations[lang].theme}: {colorTheme.name[lang]}
              </button>
              <div className="GalleryPage-dropdown-content">
                <button onClick={() => setColorTheme(ColorTheme.theme1)}>
                  {ColorTheme.theme1.name[lang]}
                </button>
                <button onClick={() => setColorTheme(ColorTheme.theme2)}>
                  {ColorTheme.theme2.name[lang]}
                </button>
                <button onClick={() => setColorTheme(ColorTheme.theme3)}>
                  {ColorTheme.theme3.name[lang]}
                </button>
              </div>
            </div>

            <div>
              <span
                className="GalleryPage-favorite-btn"
                onClick={onFavoriteClick}
              >
                <i className={`${showFavorites ? 'fas' : 'far'} fa-heart`}></i>
              </span>
            </div>
            <div>
              <button
                className={`GalleryPage-complementary-btn ${
                  complementaryTheme
                    ? 'GalleryPage-complementary-btn-fancy'
                    : ''
                }`}
                onClick={onComplementaryToggle}
              >
                {translations[lang].fancyMode}:{' '}
                {complementaryTheme
                  ? translations[lang].on
                  : translations[lang].off}
              </button>
            </div>
            <div>
              <p className="text-center GalleryPage-figure-amount-meta">
                {translations[lang].figAmount}:
              </p>
              <input
                id="typeinp"
                type="range"
                className="GalleryPage-slider"
                min={5}
                max={25}
                defaultValue={figureAmount}
                step={5}
                onMouseUp={updateFigureAmount}
                onTouchEnd={updateFigureAmount}
              />
              <span className="GalleryPage-figure-amount"> {figureAmount}</span>
            </div>
          </div>

          <div className="GalleryPage-container">
            {poems ? (
              (showFavorites
                ? favoritePoems.filter((poem) => poem.isFavorite)
                : poems
              ).map((poemObj, index) => (
                <GalleryItem
                  key={poemObj.title}
                  poem={poemObj}
                  colorTheme={colorTheme}
                  complementaryTheme={complementaryTheme}
                  amount={figureAmount}
                  figureType={null}
                  onChange={addFavoritePoem}
                  mp3={mp3s[index]}
                />
              ))
            ) : (
              <p>nothing here</p>
            )}
            {showFavorites ? (
              favoritePoems.filter((poem) => poem.isFavorite).length >
              0 ? null : (
                <h3>You don't have any favorites yet!</h3>
              )
            ) : null}
          </div>
        </div>
      )}
      );
    </div>
  );
};

export default GalleryPage;
