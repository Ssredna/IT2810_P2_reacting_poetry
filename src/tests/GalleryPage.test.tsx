import React from 'react';
import renderer from 'react-test-renderer';
import GalleryPage from '../components/GalleryPage';
import GlobalLangContext from '../utills/GlobalLangContext';
import { Languages } from '../utills/GalleryTheme';

test('it renders correctly', () => {
  const tree = renderer
    .create(
      <GlobalLangContext.Provider value={Languages.EN}>
        <GalleryPage />
      </GlobalLangContext.Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
