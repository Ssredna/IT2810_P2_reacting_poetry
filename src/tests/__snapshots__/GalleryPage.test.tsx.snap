// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
<div>
  <div
    className="General-container"
  >
    <h1
      className="GalleryPage-title"
    >
      Gallery
    </h1>
    <div
      className="GalleryPage-actions"
    >
      <div
        className="GalleryPage-dropdown"
      >
        <button
          className="GalleryPage-dropbtn"
          style={
            Object {
              "background": "#fcc5c5",
            }
          }
        >
          Theme
          : 
          Pink
        </button>
        <div
          className="GalleryPage-dropdown-content"
        >
          <button
            onClick={[Function]}
          >
            Pink
          </button>
          <button
            onClick={[Function]}
          >
            Green
          </button>
          <button
            onClick={[Function]}
          >
            Blue
          </button>
        </div>
      </div>
      <div>
        <span
          className="GalleryPage-favorite-btn"
          onClick={[Function]}
        >
          <i
            className="far fa-heart"
          />
        </span>
      </div>
      <div>
        <button
          className="GalleryPage-complementary-btn "
          onClick={[Function]}
        >
          Fancymode
          :
           
          off
        </button>
      </div>
      <div>
        <p
          className="text-center GalleryPage-figure-amount-meta"
        >
          Number of figures
          :
        </p>
        <input
          className="GalleryPage-slider"
          defaultValue={5}
          max={25}
          min={5}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          step={5}
          type="range"
        />
        <span
          className="GalleryPage-figure-amount"
        >
           
          5
        </span>
      </div>
    </div>
    <div
      className="GalleryPage-container"
    />
  </div>
  );
</div>
`;
